middle_end/flambda2/from_lambda/closure_conversion.ml

```diff
@@ -680,15 +680,18 @@ let close_let acc env id user_visible defining_expr
             List.map (Env.find_value_approximation body_env) fields
             |> Array.of_list
           in
-          Env.add_block_approximation body_env (Name.var var) approxs
+          Some (Env.add_block_approximation body_env (Name.var var) approxs)
         | Prim (Binary (Block_load _, block, field), _) -> begin
           match Env.find_value_approximation body_env block with
-          | Value_unknown -> body_env
+          | Value_unknown -> Some body_env
           | Closure_approximation _ ->
-            Misc.fatal_errorf
-              "Closure approximation found when block approximation was \
-               expected in [Closure_conversion]: %a"
-              Named.print defining_expr
+            if Flambda_features.check_invariants ()
```

Comment:
Could you leave a CR comment documenting that this case can occur on correct code (using lazy or GADTs) and that we should turn this into a disabled-by-default warning at some point ?